### PR TITLE
[PyTorch] Fix bug with PR 2677

### PR DESCRIPTION
**Pull request by sudhakarsingh27**
# Description

https://github.com/NVIDIA/TransformerEngine/pull/2677 didn't run L1 tests and the downstream CIs failed since there was a bug where padding tokens were not ignored while returning max_logit. This PR fixes it. 

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- When THD and `return_max_logit=True`, ignore the padding tokens correctly before calculating `max_logit`

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 1]
```diff
@@ -363,13 +363,35 @@ def fused_attn_fwd(
...
+            elif max_tensor.ndim == 3:
+                assert (
+                    cu_seqlens_q_padded is not None
+                ), "cu_seqlens_q_padded must be not None when qkv_format = thd!"
+                # For THD on newer cuDNN runtimes (non-sm120), Max is [tq, h, 1] with
+                # padded positions containing junk. Mask them out with -inf.
+                actual_seqlens = cu_seqlens_q[1:] - cu_seqlens_q[:-1]
+                padded_seqlens = cu_seqlens_q_padded[1:] - cu_seqlens_q_padded[:-1]
+                pad_lens = padded_seqlens - actual_seqlens
+                b = pad_lens.shape[0]
```
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py]: fixed

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 2]
```diff
@@ -363,13 +363,35 @@ def fused_attn_fwd(
...
+                seqlens_q = (cu_seqlens_q[1:] - cu_seqlens_q[:-1]).to(device=max_tensor.device)
+                sq_idx = torch.arange(max_tensor.shape[2], device=max_tensor.device).view(
+                    1, 1, -1, 1
+                )
+                valid = sq_idx < seqlens_q.view(-1, 1, 1, 1)
+                max_tensor = max_tensor.masked_fill(~valid, float("-inf"))
+            elif max_tensor.ndim == 3:
+                assert (
+                    cu_seqlens_q_padded is not None
+                ), "cu_seqlens_q_padded must be not None when qkv_format = thd!"
```
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py]: fixed

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 3]
```diff
@@ -363,13 +363,37 @@ def fused_attn_fwd(
...
+                seqlens_q = (cu_seqlens_q[1:] - cu_seqlens_q[:-1]).to(device=max_tensor.device)
+                sq_idx = torch.arange(max_tensor.shape[2], device=max_tensor.device).view(
+                    1, 1, -1, 1
+                )
+                valid = sq_idx < seqlens_q.view(-1, 1, 1, 1)
+                max_tensor = max_tensor.masked_fill(~valid, float("-inf"))
+            elif max_tensor.ndim == 3:
+                if cu_seqlens_q_padded is not None:
+                    # For THD on newer cuDNN runtimes (non-sm120), Max is [tq, h, 1] with
+                    # padded positions containing junk. Mask them out with -inf.
```
Comment (cyanguwa) [inline transformer_engine/pytorch/cpp_extensions/fused_attn.py]: Maybe point out that this is for THD + pad_between_seqs=True + non-sm120 + cuDNN>9.6 situations?

File: transformer_engine/pytorch/cpp_extensions/fused_attn.py [hunk 4]
```diff
@@ -363,13 +363,37 @@ def fused_attn_fwd(
...
+                    b = pad_lens.shape[0]
+
+                    # Stack [actual, pad] per batch into counts: e.g. [3,1, 3,1, 2,2, 7,1]
+                    counts = torch.stack([actual_seqlens, pad_lens], dim=1).flatten()
+                    # Tile [T, F] per sequence: [T,F, T,F, T,F, T,F]
+                    values = torch.tensor([True, False], device=max_tensor.device).repeat(b)
+                    # Expand: T×3, F×1, T×3, F×1, T×2, F×2, T×7, F×1 → TTTF|TTTF|TTFF|TTTTTTTF
+                    valid = torch.repeat_interleave(values, counts)
+                    # Finally, replace invalid (F) positions with -inf
+                    max_tensor = max_tensor.masked_fill(~valid.view(-1, 1, 1), float("-inf"))
```
Comment (cyanguwa) [inline transformer_engine/pytorch/cpp_extensions/fused_attn.py:397]: I feel the logic (both in ndim=3 branch and ndim=4) is a bit convoluted - it would generate quite a few small kernels. I don't know if there's a more elegant way to program this, or simply zeroing out the entire Max at initialization would be more efficient performance-wise. I'm going to approve for the functionality given that this is also kind of a niche case (THD + return_max_logit=True + pad_between_seqs=True).

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/cpp_extensions/fused_attn.py:397]: Do we have control over `stats/max` tensors? I thought they're passed over by cuDNN.
